tests/keras/layers/test_convolutional.py

```diff
@@ -61,27 +61,32 @@ def test_convolution_2d():
 
     for border_mode in ['valid', 'same']:
         for subsample in [(1, 1), (2, 2)]:
-            if border_mode == 'same' and subsample != (1, 1):
-                continue
-
-            layer_test(convolutional.Convolution2D,
-                       kwargs={'nb_filter': nb_filter,
-                               'nb_row': 3,
-                               'nb_col': 3,
-                               'border_mode': border_mode,
-                               'subsample': subsample},
-                       input_shape=(nb_samples, stack_size, nb_row, nb_col))
-
-            layer_test(convolutional.Convolution2D,
-                       kwargs={'nb_filter': nb_filter,
-                               'nb_row': 3,
-                               'nb_col': 3,
-                               'border_mode': border_mode,
-                               'W_regularizer': 'l2',
-                               'b_regularizer': 'l2',
-                               'activity_regularizer': 'activity_l2',
-                               'subsample': subsample},
-                       input_shape=(nb_samples, stack_size, nb_row, nb_col))
+            for filter_dilation in [(1, 1), (2, 2)]:
+                if border_mode == 'same' and subsample != (1, 1):
+                    continue
+                if subsample != (1, 1) and filter_dilation != (1, 1):
```

Comment:
I see, so, the test is either testing filter_dilation or subsampling.

This test logic is getting a bit complicated. Perhaps it is worth considering extracting the layer_test calls out into a separate function and splitting up the actual test cases into separate test functions? This would also make the test output more readable as there could be more descriptive test names for each case.

Also, in conv_output_length(), subsample and filter_dilation interact via output_length. I can see from the code that this isn't likely to cause issues, but this isn't tested, so any future changes that might disturb that interaction ought to also have a test. This could be done by just removing these two lines.